core/src/upgrade/apply.rs

```diff
@@ -105,6 +108,7 @@ where
     },
     Upgrade {
         future: Pin<Box<U::Future>>,
+        name: SmallVec<[u8; 32]>,
```

Comment:
A stream upgrade involves various I/O operations on many layers. I don't think a heap allocated 32 byte vector have a significant performance impact.

Thus, unless the performance improvement is proven through a benchmark, I suggest allocating on the heap here.

```suggestion
        name: Vec<u8>,
```

Reply:
Given that all protocol names I’m aware of are quite small, I see no reason to unconditionally allocate on the heap — why should that be the default that needs no reason? I could reduce the inline capacity to keep `size_of` equal to that of a `Vec<u8>`, if you want.

Reply:
I am with @rkuhn here. I think it is good to avoid allocations for the majority of the cases. I don't see the use of a different type as too much complexity.

Reply:
> I see no reason to unconditionally allocate on the heap — why should that be the default that needs no reason?

Reason for me to not use `SmallVec` by default is that it adds noise, additional compile time and increases the attack vector in our dependency supply chain. Though the last one might be a lost fight in the first place.

I wrote a benchmark to try to replicate a performance benefit using `SmallVec` here. I have not been able thus far. Please take a look at https://github.com/Actyx/rust-libp2p/pull/95. Maybe I can nerd-snipe either of you into proving me wrong @rkuhn and @thomaseizinger.

I am sorry in case this comes across as nit-picky. I am not trying to start an argument. I am genuinely interested in performance optimizations in rust-libp2p. I am curious in which cases we could optimize through stack allocation. I don't think this is one of those cases.

On a more general note, again out of curiosity, is either of you aware of a case where heap allocation **is** faster than stack allocation? I can not come up with a real world case where the heap would be faster than the stack.

Reply:
To me this is foremost a question of using the most appropriate data structure for the job. SmallVec was already in the dependencies, and it is better suited than Vec due to usage characteristics, so I used it.

---

Regarding the benchmark: memory management cannot be benchmarked in this fashion. In general, microbenchmarks like yours are very rarely meaningful and even then they require careful interpretation of the result. The reason is that most of the time you are looking at a huge machine doing a tiny task, with all sorts of optimisations kicking in. And when you run a relevant task instead, different optimisations will completely change the picture. It is quite obvious that a stack allocation takes less than one CPU cycle and that a heap allocation can in general not be this fast. If your microbenchmark finds them to be equally fast, then your heap usage allows the `malloc` implementation to optimise down to a single pointer subtraction — which is a very likely thread-local optimisation to make for `malloc()`.

But all of this theory is irrelevant in my opinion because we’re not optimising for the last percent of performance (as far as I’m aware we’re nowhere close to getting there at this point). So I just propose usage of the most appropriate data structure at hand.

Reply:
I think it is reasonable to trust in CS knowledge like "stack allocations are faster than heap", even if you can't produce a micro benchmark that proves it.

Soon, this will be replaced by `Cow` so I am happy either way. If @mxinden feels strongly, let's go with `Vec`.

Reply:
BTW: this discussion uses “heap” and “stack”, but “stack” doesn’t actually enter the picture in the code in question since we’re talking about a memory slot in a generator, which will end up spawned onto a thread pool, so allocated on the heap in any case.

Reply:
Right, what would be the right terminology then (out of curiosity?). "local variable"?

Reply:
That sounds appropriate to me — it’ll end up being stored in a compiler-generated `enum` variant if it lives across an `.await` point.

Reply:
Thanks for expanding here. I think the points above are valid. That said, I haven't yet formed a good guiding principle for rust-libp2p.

As said above, the concrete instance is not relevant. Let's get this merged. Looking forward to the easier debugging of protocol upgrades on streams.